c/tskit/tables.c

```diff
@@ -8105,6 +8105,285 @@ tsk_table_collection_subset(
+    tsk_id_t *node_mapping, tsk_size_t other_num_nodes, tsk_size_t num_shared_nodes)
+{
+    int ret = 0;
+    tsk_id_t k, i;
+    tsk_id_t *self_nodes = NULL;
+    tsk_id_t *other_nodes = NULL;
+    tsk_table_collection_t self_copy;
+    tsk_table_collection_t other_copy;
+
+    self_nodes = malloc(num_shared_nodes * sizeof(*self_nodes));
+    other_nodes = malloc(num_shared_nodes * sizeof(*other_nodes));
+    if (self_nodes == NULL || other_nodes == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    ret = tsk_table_collection_init(&self_copy, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_init(&other_copy, 0);
+    if (ret != 0) {
+        goto out;
+    }
+
+    i = 0;
+    for (k = 0; k < (tsk_id_t) other_num_nodes; k++) {
+        if (node_mapping[k] != TSK_NULL) {
+            self_nodes[i] = node_mapping[k];
+            other_nodes[i] = k;
+            i++;
+        }
+    }
+
+    // TODO: strict sort before checking equality
+    ret = tsk_table_collection_copy(self, &self_copy, TSK_NO_INIT);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_copy(other, &other_copy, TSK_NO_INIT);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_provenance_table_clear(&other_copy.provenances);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_provenance_table_clear(&self_copy.provenances);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_subset(&self_copy, self_nodes, num_shared_nodes);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_subset(&other_copy, other_nodes, num_shared_nodes);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = !tsk_table_collection_equals(&self_copy, &other_copy);
+    if (ret != 0) {
+        ret = TSK_ERR_DIFF_HISTORIES;
+        goto out;
+    }
+
+out:
+    tsk_table_collection_free(&self_copy);
+    tsk_table_collection_free(&other_copy);
+    tsk_safe_free(other_nodes);
+    tsk_safe_free(self_nodes);
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_table_collection_union(tsk_table_collection_t *self, tsk_table_collection_t *other,
+    tsk_id_t *node_mapping, tsk_size_t num_nodes, tsk_flags_t options)
+{
+    int ret = 0;
+    tsk_id_t k, i, new_ind, new_pop, new_parent, new_child;
+    tsk_size_t num_shared_nodes = 0;
+    tsk_node_t node;
+    tsk_individual_t ind;
+    tsk_population_t pop;
+    tsk_edge_t edge;
+    tsk_mutation_t mut;
+    tsk_site_t site;
+    tsk_id_t *node_map = NULL;
+    tsk_id_t *individual_map = NULL;
+    tsk_id_t *population_map = NULL;
+    tsk_id_t *site_map = NULL;
+    bool add_populations = !(options & TSK_UNION_NO_ADD_POP);
+    bool check_shared_portion = !(options & TSK_UNION_NO_CHECK_SHARED);
+
+    /* node_mapping relates id in other (index) to equivalent node in self or
+     * TSK_NULL if not present */
+    if (num_nodes != (tsk_size_t) other->nodes.num_rows) {
```

Comment:
I don't think there's any point in having this parameter if it must always be ``other->nodes.num_rows``. It doesn't make things any safer because the user can still provide a buffer the wrong size.

Reply:
yes, I was thinking it  would be more useful for calls from the Python side, in which case the buffer would always equal the size of the array. Should I deal with this in `_tskitmodule.c` then?

Reply:
ok, changed to check for this in the py-c interface.